components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt

```diff
@@ -98,9 +101,19 @@ class FxaWebChannelFeature(
      */
     private class WebChannelViewContentMessageHandler(
         private val accountManager: FxaAccountManager,
+        private val serverConfig: ServerConfig,
         private val fxaCapabilities: Set<FxaCapability>
     ) : MessageHandler {
+        @SuppressWarnings("ComplexMethod")
         override fun onPortMessage(message: Any, port: Port) {
+            val senderOrigin = URL(port.senderUrl()).host
+            val expectedOrigin = URL(serverConfig.contentUrl).host
```

Comment:
I'm not super familiar with `java.net.URL` but a quick read of the docs suggests that `host` may not be sufficient here, because it may not include e.g. the port number or URL scheme. I've seen enough origin-matching bugs in my time to be very wary of getting this right.  I'd like to see this check factored out into some sort of `isSameOrigin()` function and have tests for at least the following cases:

* `isSameOrigin("https://example.com", https://example.com") -> true` (simple match)
* `isSameOrigin("https://example.com/foo", https://example.com/bar") -> true` (matches ignoring path)
* `isSameOrigin("https://example.com", https://example.com:443") -> true` (default port for scheme)
* `isSameOrigin("https://example.com:8000", https://example.com:8000") -> true` (non-default port match)
* `isSameOrigin("https://example.com", https://example.net") -> false` (host mismatch)
* `isSameOrigin("https://example.com", http://example.com") -> false` (scheme mismatch)
* `isSameOrigin("https://example.com", https://example.com:444") -> false` (non-default port mismatch)
* `isSameOrigin("https://example.com:444", https://example.com:445") -> false` (explicit port mismatch)
* `isSameOrigin("https://example.com", http://example.com:443") -> false` (scheme mismatch despite explicit port)
* `isSameOrigin("https://example.com:8000", http://example.com:8000") -> true` (scheme mismatch despite non-default port)

I had a bit of a look around but wasn't able to find an easy way to do this ([this package](https://kotlinlang.org/api/latest/jvm/stdlib/org.w3c.dom.url/) provides a `URL` class with an `origin` property, but looks like it's only when targetting javascript?). Absent something better, I suggest something like (pardon my pseudocode):

```
def canonicalizeOrigin(v):
  u = URL(v)
  c = URL(u.protocol, u.host, u.port == -1 ? u.defaultPort : u.port, "")
  return c.toString()

def isSameOrigin(v1, v2):
    return canonicalizeOrigin(v1) == canonicalizeOrigin(v2)
```

It's not very efficient, but it's easy to understand.

I also think it would be valuable to add some belt-and-braces security checks for things we would never expect to se in the wild:

* Reject any messages received from a non-https non-localhost URL
* Reject any messages received with from a URL with non-empty userinfo component